File: Source/Fantasy.Jobs/Resources/GlobalMutexService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.ServiceModel;
using Fantasy.Jobs.Scheduling;
using Fantasy.Jobs.Management;

namespace Fantasy.Jobs.Resources
{
    [ServiceBehavior(InstanceContextMode=InstanceContextMode.Single, ConcurrencyMode=ConcurrencyMode.Multiple, Namespace=Consts.JobServiceNamespaceURI)] 
    public class GlobalMutexService : WCFSingletonService, IGlobalMutexService
    {

        private List<Mutex> _allocated = new List<Mutex>();
        private object _syncRoot = new object();
        private IScheduleService _scheduleService;
        private IJobDispatcher _jobDispatcher;

        public override void InitializeService()
        {
            _scheduleService = this.Site.GetRequiredService<IScheduleService>();
            _jobDispatcher = this.Site.GetService<IJobDispatcher>();
            base.InitializeService();
        }

        private class Mutex
        {
            public string Name { get; set; }

            public long Cookie { get; set; }
        }


        #region IGlobalMutexService Members

        public bool IsAvaiable(string key)
        {
            lock (_syncRoot)
            {
                return ! this._allocated.Any(m => String.Equals(key, m.Name, StringComparison.OrdinalIgnoreCase)); 
            }
        }

        public bool Request(string key, TimeSpan timeout)
        {
            lock (_syncRoot)
            {
                Mutex mutex =  this._allocated.Find(m => String.Equals(key, m.Name, StringComparison.OrdinalIgnoreCase));
                if (mutex == null)
                {
                    mutex = new Mutex() { Name = key };
                    mutex.Cookie = _scheduleService.Register(DateTime.Now + timeout, () => { this.Release(mutex); });
                    this._allocated.Add(mutex);
                    return true;
                }
                else
                {
                    return false;
                }
            }
        }

        private void Release(Mutex mutex)
        {
            lock (_syncRoot)
            {
                _allocated.Remove(mutex);
            }
            if (_jobDispatcher != null)
            {
                _jobDispatcher.TryDispatch();
            }

        }

        public void Release(string key)
        {
            lock (_syncRoot)
            {
                Mutex mutex = this._allocated.Find(m => String.Equals(key, m.Name, StringComparison.OrdinalIgnoreCase));
                if (mutex != null)
                {
                    this.Release(mutex);
                    _scheduleService.Unregister(mutex.Cookie);
                }
            }
        }

        #endregion
    }
}

File: Fantasy.Jobs/src/fantasy/jobs/resources/GlobalMutexService.java
package fantasy.jobs.resources;

import Fantasy.Jobs.Scheduling.*;
import Fantasy.Jobs.Management.*;

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
//[ServiceBehavior(InstanceContextMode=InstanceContextMode.Single, ConcurrencyMode=ConcurrencyMode.Multiple, Namespace=Consts.JobServiceNamespaceURI)]
public class GlobalMutexService extends WCFSingletonService implements IGlobalMutexService
{

	private java.util.ArrayList<Mutex> _allocated = new java.util.ArrayList<Mutex>();
	private Object _syncRoot = new Object();
	private IScheduleService _scheduleService;
	private IJobDispatcher _jobDispatcher;

	@Override
	public void InitializeService()
	{
		_scheduleService = this.Site.<IScheduleService>GetRequiredService();
		_jobDispatcher = this.Site.<IJobDispatcher>GetService();
		super.InitializeService();
	}

	private static class Mutex
	{
		private String privateName;
		public final String getName()
		{
			return privateName;
		}
		public final void setName(String value)
		{
			privateName = value;
		}

		private long privateCookie;
		public final long getCookie()
		{
			return privateCookie;
		}
		public final void setCookie(long value)
		{
			privateCookie = value;
		}
	}


//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#region IGlobalMutexService Members

	public final boolean IsAvaiable(String key)
	{
		synchronized (_syncRoot)
		{
//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
			return ! this._allocated.Any(m => String.equals(key, m.getName(), StringComparison.OrdinalIgnoreCase));
		}
	}

	public final boolean Request(String key, TimeSpan timeout)
	{
		synchronized (_syncRoot)
		{
//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
			Mutex mutex = this._allocated.Find(m => String.equals(key, m.getName(), StringComparison.OrdinalIgnoreCase));
			if (mutex == null)
			{
				Mutex tempVar = new Mutex();
				tempVar.setName(key);
				mutex = tempVar;
//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
				mutex.setCookie(_scheduleService.Register(new java.util.Date() + timeout, () =>)
				{
					this.Release(mutex);
				}
			   );
				this._allocated.add(mutex);
				return true;
			}
			else
			{
				return false;
			}
		}
	}

	private void Release(Mutex mutex)
	{
		synchronized (_syncRoot)
		{
			_allocated.remove(mutex);
		}
		if (_jobDispatcher != null)
		{
			_jobDispatcher.TryDispatch();
		}

	}

	public final void Release(String key)
	{
		synchronized (_syncRoot)
		{
//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
			Mutex mutex = this._allocated.Find(m => String.equals(key, m.getName(), StringComparison.OrdinalIgnoreCase));
			if (mutex != null)
			{
				this.Release(mutex);
				_scheduleService.Unregister(mutex.getCookie());
			}
		}
	}

//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#endregion
}